vivisect/analysis/i386/golang.py

```diff
@@ -0,0 +1,202 @@
+'''
+Perform further analysis on 32-bit GO language executables.
+
+This module extends code flow analysis to runtime_main for Windows PE binaries.
+GO binaries start from a single export and proceed thru several functions
+that initialize GO.  Specific application code is launched from the GO function
+runtime_main(), which is invoked by "call eax" with an address placed on the
+stack many calls earlier.  Vivisect code flow analysis does not track the
+address; this module finds the address and invokes makeFunction(va).
+
+Samples have been found with different instruction sequences for reaching the
+address of runtime_main(); this module attempts all observed sequences.
+'''
+
+import envi
+import envi.archs.i386.disasm
+
+
+def analyze(vw):
+    '''
+    Perform further analysis on GO language executables.
+    '''
+    # Make sure it is a PE file, with "Go build ID:" in the first few bytes
+    # of the .text segment (versus a .upxN segment of a packed sample).
+    if (vw.getMeta('Format') != 'pe') and \
+       (vw.getMeta('Architecture') != 'i386'):
+        return
+    has_go_build = False
+    for segment in vw.getSegments():
+        va, size, name, filename = segment
+        if name == '.text':
+            bytez = vw.readMemory(va, 10000)
```

Comment:
Is the build id string guaranteed to be in the first 10k bytes, or would it help to just pass in the segment size here?

Reply:
No guarantees, but all examples are in the first 4k.  I use 10k as a safety margin.  The segment can be huge so the idea is to save time.

Reply:
what happens if the memory map is less than 10k past the start of .text?
i realize this is uncommon, but could cause problems since this is the first test beyond PE and architecture.

Reply:
nm.  Viv polite returns the amount of data that exists in the memory map from that point.